opencti-platform/opencti-graphql/src/domain/backgroundTask-common.js

```diff
@@ -1,38 +1,104 @@
 export const TASK_TYPE_LIST = 'LIST';
+
+export const ACTION_TYPE_DELETE = 'DELETE';
 export const ACTION_TYPE_SHARE = 'SHARE';
 export const ACTION_TYPE_UNSHARE = 'UNSHARE';
 
-export const checkActionValidity = (user, actions) => {
-  const askForDeletion = actions.filter((a) => a.type === ACTION_TYPE_DELETE).length > 0;
-  if (askForDeletion) {
-    // If deletion action available, user need to have the right capability
-    const userCapabilities = R.flatten(user.capabilities.map((c) => c.name.split('_')));
-    const isAuthorized = userCapabilities.includes(BYPASS) || userCapabilities.includes(KNOWLEDGE_DELETE);
+const areParentTypesKnowledge = (parentTypes) => parentTypes && parentTypes.flat().every((type) => isKnowledge(type));
+
+// check a user has the right to create a list or a query background task
+export const checkActionValidity = async (context, user, input, scope, taskType) => {
+  const { actions, filters, ids } = input;
+  const userCapabilities = R.flatten(user.capabilities.map((c) => c.name.split('_')));
+  if (scope === 'KNOWLEDGE') { // 01. Background task of scope Knowledge
+    // 1.1. The user should have the capability KNOWLEDGE_UPDATE
+    const isAuthorized = userCapabilities.includes(BYPASS) || userCapabilities.includes(KNOWLEDGE_UPDATE);
     if (!isAuthorized) {
       throw ForbiddenAccess();
     }
+    const askForDeletion = actions.filter((a) => a.type === ACTION_TYPE_DELETE).length > 0;
+    if (askForDeletion) {
+      // 1.2. If deletion action available, the user should have the capability KNOWLEDGE_DELETE
+      const isDeletionAuthorized = userCapabilities.includes(BYPASS) || userCapabilities.includes(KNOWLEDGE_DELETE);
+      if (!isDeletionAuthorized) {
+        throw ForbiddenAccess();
+      }
+    }
+    // 1.3. Check the modified entities are of type Knowledge
+    if (taskType === TASK_TYPE_QUERY) {
+      const parentTypes = JSON.parse(filters).entity_type?.map((n) => getParentTypes(n.id));
+      const isNotKnowledges = !areParentTypesKnowledge(parentTypes);
+      if (isNotKnowledges) {
+        throw ForbiddenAccess(undefined, 'The targeted ids are not knowledges.');
+      }
+    } else if (taskType === TASK_TYPE_LIST) {
+      const objects = await Promise.all(ids.map((id) => storeLoadById(context, user, id, ABSTRACT_STIX_OBJECT)));
+      const isNotKnowledges = objects.includes(undefined)
+        || !areParentTypesKnowledge(objects.map((o) => o.parent_types));
+      if (isNotKnowledges) {
+        throw ForbiddenAccess(undefined, 'The targeted ids are not knowledges.');
+      }
+    } else {
+      throw Error('A background task should be of type query or list.');
+    }
+  } else if (scope === 'USER') { // 02. Background task of scope Notification
+    // Check the modified entities are Notifications
+    // and the user has the right to modify them (= notifications are the ones of the user OR the user has SET_ACCESS capability)
+    if (taskType === TASK_TYPE_QUERY) {
```

Comment:
Can we have the same order as previous ?
first check right
second check types

Reply:
No we can't because for taskType = TASK_TYPE_LIST, the order is compulsory (types then rights) because the rights checks used the fact that the objects are of type notification